amethyst_rendy/shaders/fragment/header/ltc.glsl

```diff
@@ -0,0 +1,540 @@
+// "How to solve a cubic equation, revisited"
+// http://momentsingraphics.de/?p=105
+vec3 solve_cubic(vec4 Coefficient)
+{
+    // Normalize the polynomial
+    Coefficient.xyz /= Coefficient.w;
+    // Divide middle coefficients by three
+    Coefficient.yz /= 3.0;
+
+    float A = Coefficient.w;
+    float B = Coefficient.z;
+    float C = Coefficient.y;
+    float D = Coefficient.x;
+
+    // Compute the Hessian and the discriminant
+    vec3 Delta = vec3(
+        -Coefficient.z*Coefficient.z + Coefficient.y,
+        -Coefficient.y*Coefficient.z + Coefficient.x,
+        dot(vec2(Coefficient.z, -Coefficient.y), Coefficient.xy)
+    );
+    // In contrast to the reference implementation we add 0.0001 
+    // to the discriminant to avoid some unstable results.
+    float Discriminant = dot(vec2(4.0*Delta.x, -Delta.y), Delta.zy)+0.0001;
+
+    vec3 RootsA, RootsD;
+
+    vec2 xlc, xsc;
+
+    // Algorithm A
+    {
+        float A_a = 1.0;
+        float C_a = Delta.x;
+        float D_a = -2.0*B*Delta.x + Delta.y;
+
+        // Take the cubic root of a normalized complex number
+        float Theta = atan(sqrt(Discriminant), -D_a)/3.0;
+
+        float x_1a = 2.0*sqrt(-C_a)*cos(Theta);
+        float x_3a = 2.0*sqrt(-C_a)*cos(Theta + (2.0/3.0)*PI);
+
+        float xl;
+        if ((x_1a + x_3a) > 2.0*B)
+            xl = x_1a;
+        else
+            xl = x_3a;
+
+        xlc = vec2(xl - B, A);
+    }
+
+    // Algorithm D
+    {
+        float A_d = D;
+        float C_d = Delta.z;
+        float D_d = -D*Delta.y + 2.0*C*Delta.z;
+
+        // Take the cubic root of a normalized complex number
+        float Theta = atan(D*sqrt(Discriminant), -D_d)/3.0;
+
+        float x_1d = 2.0*sqrt(-C_d)*cos(Theta);
+        float x_3d = 2.0*sqrt(-C_d)*cos(Theta + (2.0/3.0)*PI);
+
+        float xs;
+        if (x_1d + x_3d < 2.0*C)
+            xs = x_1d;
+        else
+            xs = x_3d;
+
+        xsc = vec2(-D, xs + C);
+    }
+
+    float E =  xlc.y*xsc.y;
+    float F = -xlc.x*xsc.y - xlc.y*xsc.x;
+    float G =  xlc.x*xsc.x;
+
+    vec2 xmc = vec2(C*F - B*G, -B*F + C*E);
+
+    vec3 Root = vec3(xsc.x/xsc.y, xmc.x/xmc.y, xlc.x/xlc.y);
+
+    if (Root.x < Root.y && Root.x < Root.z)
+        Root.xyz = Root.yxz;
+    else if (Root.z < Root.x && Root.z < Root.y)
+        Root.xyz = Root.xzy;
+
+    return Root;
+}
+
+
+/*
+ * Get intensity of light from the arealight given by `points` at the point `P`
+ * with normal `N` when viewed from direction `P`.
+ * @param N Normal
+ * @param V View Direction
+ * @param P Vertex Position
+ * @param Minv Matrix to transform from BRDF distribution to clamped cosine distribution
+ * @param points Light quad vertices
+ * @param twoSided Whether the light is two sided
+ */
+vec3 ltc_evaluate_ellipse(
+    vec3 N, 
+    vec3 V, 
+    vec3 P, 
+    float NdotV,
+    mat3 Minv, 
+    vec3 points[4], 
+    bool two_sided,
+    bool sphere
+)
+{
+
+
+    if (sphere) {
```

Comment:
Can we somehow get this to be non-branching, maybe higher up where it is called this can be static? or perhaps we need a seperate shader? something to discuss to avoid it. This probably ties into a higher discussion about how to handle `#ifdef` for shader compilation

Reply:
I tried to avoid too many different light structs and thus shader bindings. Not sure what a sane limit is here. 
Not sure how bad that branching is, as most of the fragments on screen will take the same branch.